### PR TITLE
add @once - forces a method to run only one time

### PR DESCRIPTION
**Pull request by zivl**
After first run, the return value is saved and for every other run of the method it will return the first result right away

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ I *highly* recommend against using that globals build as it's quite strange you'
 * [@debounce](#debounce)
 * [@throttle](#throttle)
 * [@time](#time)
+* [@once](#once)
```
Comment (nrdobie) [inline README.md]: Should add `:new:` after this just to let people know it was just added.

File: src/once.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  if (typeof fn !== 'function') {
+    throw new SyntaxError(`@once can only be used on functions, not: ${fn}`);
+  }
+
+  let returnedValue;
+
+  return {
+    ...descriptor,
+    value() {
+      if (!returnedValue) {
```
Comment (nrdobie) [inline src/once.js]: Should use a second boolean variable to check if the function has been run. If the first run of the function returns a falsy value, it allows the function to run multiple times. An example would be a function that has no return value would always run.

**Comment by jayphelps** [comment]
Thanks for the PR!

I'm hesitant to add this decorator because I'm not confident it will see much usage--I don't want core-decorators to include _every_ decorator, especially those which overlap lodash.

e.g. you can use the `@decorate` decorator to apply most lodash utilities

```js
class Foo {
  @decorate(_.once)
  runSumOnlyOnce(num1, num2) {
    return num1 + num2;
  }
}
```

There are some existing decorators that overlap with lodash, but mostly I felt they were more likely to see use or they were added prior to the addition of `@decorate`.

I'm open to discussion of course! 😄 Thoughts?

File: src/once.js (hunk 2 above)
Comment (zivl) [inline src/once.js]: sure! missed that.. thanks!

File: README.md (hunk 1 above)
Comment (zivl) [inline README.md]: ok, right.. 
should I remove older "new" labels? what's the `readme` routine here?

**Comment by zivl** [comment]
@jayphelps - actually I haven't notice `lodash` has `once` function... yet, I agree, no need to re-write `lodash` in decorators. Never the less, I think `once` capability should be part of `core-decorators` project. Also, thinking out loud, looks like with `ES7` - all `lodash` utilities that are related to functions - should be transformed into decorators - this is how I see it.

**Comment by jayphelps** [comment]
@zivl 

> Also, thinking out loud, looks like with ES7 - all lodash utilities that are related to functions - should be transformed into decorators - this is how I see it.

Decorators aren't in ES7 aka ES2016 (they aren't in any finalized spec yet) but I get your sentiment. I _really_ don't want to reinvent them because we'll have to deal with the same bugs they worked through. Instead, if someone wants to, I think a library could simply import `lodash` and provided decorator versions of them that just defer to the lodash implementations.

I'm not sure core-decorators is the place for that--maybe it is. I'd be interested to hear what the community thinks.

**Comment by jayphelps** [comment]
Going to keep this open for a couple weeks to see if someone has further thoughts on lodash duplication or some sort of lodash wrapper library.

**Comment by nrdobie** [comment]
I think it would make sense to create another package like `lodash-decorators`. It would be possible to have the package use the `@decorate` from `core-decorators` and import the standalone Lodash functions, e.g. `lodash.once`. Going one step further, the package could try to use the Lodash version installed by the user via `optionalDependency` to help reduce bloat.

**Comment by jayphelps** [comment]
@nrdobie that's very close to my thinking as well. I imagined you'd need to take a similar approach to lodash--providing ESM bundles so they can use Rollup/Webpack2 to tree shake.

***

Actually, **this already exists** https://www.npmjs.com/package/lodash-decorators (doesn't use core-decorators cause it really doesn't need to)

Seems like this is a solved problem?

**Comment by jayphelps** [comment]
Now that https://www.npmjs.com/package/lodash-decorators exists, I'm going to think about deprecating then removing all our decorators that overlap with it and keep core-decorators only to more fundamental ones like `@readonly` and `@autobind`

**Comment by zivl** [comment]
Hey, 
I think that a library needs to evolve. Over time, libraries that are not evolving are eventually deprecated due to *out-of-date relevance*.
So sure, there are overlaps here and there, but it is up to the consumer wether to use both libraries or only one of them (or in the `lodash` case, can use only a single or a few functions).

let me know what's the final decision so I will resolve conflicts in order to merge

**Comment by jayphelps** [comment]
@zivl I'm confused. Certainly libraries need to evolve. Why would core-decorators duplicate the effort of lodash-decorators? If not doing so ends up meaning that core-decorators is no longer relevant, I think that'd be a good thing--but I strongly disagree. They have different goals.

I'm pretty happy with deprecating then removing decorators that overlap with lodash.

I really do appreciate your work on this PR and the discussion, but I don't want to duplicate effort. We could never keep up with the robustness of lodash.

**Comment by jayphelps** [comment]
I've made a initial PR to kick start the deprecation process of the overlapping decorators: #110 feel free to comment!

**Comment by zivl** [comment]
yeah I got your point... I guess you're right.
anyway, I'm happy we could discuss that issue and come to that conclusion.

Cheers!

**Comment by jayphelps** [comment]
Thanks for understanding! 🎉

**Comment by jayphelps** [comment]
We've officially deprecated all the decorators that overlap with [lodash-decorators](https://www.npmjs.com/package/lodash-decorators), so this one won't make it. 😢  Thank you for the work you put into this!

**Pull request closed without merging**